File: 0410-split-array-largest-sum/0410-split-array-largest-sum.py
class Solution:
    def splitArray(self, nums: List[int], k: int) -> int:
        def canSplit(max_sum)->bool:
            subarrays = 1
            current_sum = 0

            for num in nums:
                if current_sum + num > max_sum:
                    subarrays += 1
                    current_sum = num
                else:
                    current_sum += num
            
            return subarrays <= k
        
        low = max(nums)
        high = sum(nums)
        ans = high

        while low <= high:
            mid = (low + high) // 2

            if canSplit(mid):
                ans = mid
                high = mid - 1
            else:
                low = mid + 1
        
        return ans

File: 0410-split-array-largest-sum/0410-split-array-largest-sum.cpp
class Solution {
private:
    bool canSplit(vector<int>&nums , int k , int max_sum){
        int subarrays = 1;
        int currentSum = 0;

        for(int num : nums){
            if(currentSum + num > max_sum){
                subarrays++;
                currentSum = num;
            }
            else{
                currentSum += num;
            }
        }

        return subarrays <= k;
    }

public:
    int splitArray(vector<int>& nums, int k) {
        int low = *max_element(nums.begin() , nums.end());
        int high = accumulate(nums.begin() , nums.end() , 0);
        int ans = high;

        while(low <= high){
            int mid = low + (high - low) / 2;

            if(canSplit(nums , k , mid)){
                ans = mid;
                high = mid - 1;
            }
            else{
                low = mid + 1;
            }
        }

        return ans;
    }
};